contracts/src/BoundlessMarket.sol

```diff
@@ -214,52 +216,78 @@ contract BoundlessMarket is
         // NOTE: Signature checks and recursive verification happen inside the assessor.
         bytes32[] memory requestDigests = new bytes32[](1);
         requestDigests[0] = fill.requestDigest;
-        bytes32 assessorJournalDigest =
-            sha256(abi.encode(AssessorJournal({requestDigests: requestDigests, root: claimDigest, prover: prover})));
+        bytes32 assessorJournalDigest = sha256(
+            abi.encode(
+                AssessorJournal({
+                    requestDigests: requestDigests,
+                    selectors: assessorFill.selectors,
+                    root: claimDigest,
+                    prover: assessorFill.prover
+                })
+            )
+        );
+        // Verification that the provided seal matches the required selector.
+        if (assessorFill.selectors.indices.length == 1 && assessorFill.selectors.values[0] != bytes4(fill.seal[0:4])) {
+            revert SelectorMismatch(assessorFill.selectors.values[0], bytes4(fill.seal[0:4]));
+        }
         // Verification of the assessor seal does not need to comply with FULFILL_MAX_GAS_FOR_VERIFY.
-        VERIFIER.verify(assessorSeal, ASSESSOR_ID, assessorJournalDigest);
+        VERIFIER.verify(assessorFill.seal, ASSESSOR_ID, assessorJournalDigest);
     }
 
     /// @inheritdoc IBoundlessMarket
-    function verifyBatchDelivery(Fulfillment[] calldata fills, bytes calldata assessorSeal, address prover)
-        public
-        view
-    {
+    function verifyBatchDelivery(Fulfillment[] calldata fills, FulfillmentAssessor calldata assessorFill) public view {
         // TODO(#242): Figure out how much the memory here is costing. If it's significant, we can do some tricks to reduce memory pressure.
         bytes32[] memory claimDigests = new bytes32[](fills.length);
         bytes32[] memory requestDigests = new bytes32[](fills.length);
-        for (uint256 i = 0; i < fills.length; i++) {
+        for (uint8 i = 0; i < fills.length; i++) {
             requestDigests[i] = fills[i].requestDigest;
             claimDigests[i] = ReceiptClaimLib.ok(fills[i].imageId, sha256(fills[i].journal)).digest();
+            // Verification that the provided seal matches the required selector.
+            for (uint8 j = 0; j < assessorFill.selectors.indices.length; j++) {
+                if (assessorFill.selectors.indices[j] == i) {
+                    if (assessorFill.selectors.values[j] != bytes4(fills[i].seal[0:4])) {
+                        revert SelectorMismatch(assessorFill.selectors.values[j], bytes4(fills[i].seal[0:4]));
+                    }
+                    break;
+                }
+            }
```

Comment:
Hmm we should really try and avoid O(n^2). Not sure which of these would be more gas efficient, but a couple ideas:

1/
Can we guarantee that the indices committed to the asssesor are sorted? If so I think we can just do something like:
```
j = 0
for (i=0; i<fills.length; i++)
    if (assessorFill.selectors.indices[j] == i) {
        // check seal
        j++;
    }
    // VERIFIER stuff
}
```


2/
We could have two loops where we first iterative fills and verifiy, then iterator assessor selectors and check those.

3/
Commit something to selectors array for every fill (some null value if theres no selector requirement)

Reply:
Also would be good to have some gas tests that test this for large batch sizes with lots of selector predicates, something similar to the Benchmark tests we have now

Reply:
Good call, I've dropped the second loop as well as added a benchmark